go-controller/pkg/cluster/ha_master.go

```diff
@@ -0,0 +1,541 @@
+	return &OvnHAMasterController{
+		kubeClient:      kubeClient,
+		cluster:         cluster,
+		ovnController:   ovnController,
+		nodeName:        nodeName,
+		manageDBServers: manageDBServers,
+		isLeader:        false,
+		leaderElector:   nil,
+	}
+}
+
+// StartHAMasterCluster runs the replication controller
+func (hacluster *OvnHAMasterController) StartHAMasterCluster() error {
+	if hacluster.manageDBServers {
+		// Always demote the OVN DBs to backup mode.
+		// After the leader election, the leader will promote the OVN Dbs
+		// to become active.
+		err := hacluster.DemoteOVNDbs(invalidIPAddress, ovnNorthDbPort, ovnSouthDbPort)
+		if err != nil {
+			// If we are not able to communicate to the OVN ovsdb-servers,
+			// then it is better to return than continue.
+			// cmd/ovnkube.go will panic if this function returns error.
+			return err
+		}
+	}
+
+	// Set up leader election process first
+	rl, err := resourcelock.New(
+		resourcelock.ConfigMapsResourceLock,
+		config.Kubernetes.OVNConfigNamespace,
+		haLeaderLockName,
+		hacluster.kubeClient.CoreV1(),
+		nil,
+		resourcelock.ResourceLockConfig{
+			Identity:      hacluster.nodeName,
+			EventRecorder: nil,
+		})
+
+	if err != nil {
+		return err
+	}
+
+	HAClusterOnStoppedLeading := func() {
+		//This node was leader and it lost the election.
+		// Whenever the node transitions from leader to follower,
+		// we need to handle the transition properly like clearing
+		// the cache. It is better to exit for now.
+		// kube will restart and this will become a follower.
+		if hacluster.manageDBServers {
+			// Stop ovn-northd and then exit.
+			_, _, _ = util.RunOVNNorthAppCtl("exit")
+		}
+		logrus.Infof("I (" + hacluster.nodeName + ") am no longer a leader. Exiting")
+		os.Exit(1)
+	}
+
+	HAClusterNewLeader := func(nodeName string) {
+		logrus.Infof(nodeName + " is the new leader")
+		wasLeader := hacluster.isLeader
+
+		if hacluster.nodeName == nodeName {
+			// Configure as leader.
+			logrus.Infof(" I (" + hacluster.nodeName + ") won the election. In active mode")
+			err = hacluster.ConfigureAsActive(nodeName)
+			if err != nil {
+				logrus.Errorf(err.Error())
+				if hacluster.manageDBServers {
+					// Stop ovn-northd before panicing.
+					_, _, _ = util.RunOVNNorthAppCtl("exit")
+				}
+				panic(err.Error())
+			}
+			hacluster.isLeader = true
+		} else if wasLeader {
+			HAClusterOnStoppedLeading()
+			// should not be reached
+			panic("This should not happen.")
+		} else {
+			// Configure as standby.
+			logrus.Infof(" I (" + hacluster.nodeName + ") lost the election. In Standby mode")
+			ep, er := hacluster.cluster.Kube.GetEndpoint(config.Kubernetes.OVNConfigNamespace, ovnkubeDbEp)
+			if er == nil {
+				er = hacluster.ConfigureAsStandby(ep)
+				if er != nil {
+					logrus.Errorf(er.Error())
+					if hacluster.manageDBServers {
+						// Stop ovn-northd and then exit
+						_, _, _ = util.RunOVNNorthAppCtl("exit")
```

Comment:
northd will be immediately restarted by kubernetes. Do we need to come up with some other way to keep it stopped?

(I would suggest touching a file somewhere in `/run` when we have leader, then we can loop waiting for that.)

Reply:
Oh, or they will try and write to a read-only database and hang?

Reply:
It will try to write to a read-only database.
So I don't think it should cause any issues.

We can do one thing - the ovn-northd init script can pause the ovn-northd after starting ovn-northd.

i.e it can run - ovs-appctl -t ovn-northd pause.

The ovnkube-master which becomes active will "resume" it.